smartapps/ady624/webcore-piston.src/webcore-piston.groovy

```diff
@@ -1761,413 +1588,457 @@ private executePhysicalCommand(rtData, device, command, params = [], delay = nul
-                        device."$command"(params as Object[])
-                        msg.m = "Executed physical command [${device.label ?: device.name}].$command($params)"
-                    }
-                } else {
-                    if (delay) { //not supported in hubitat
-                        device."$command"([delay: delay])
-                        msg.m = "Executed physical command [${device.label ?: device.name}].$command([delay: $delay])"
-                    } else {
-                        device."$command"()
-                        msg.m = "Executed physical command [${device.label ?: device.name}].$command()"
-                    }
-                }
-            }
-            if (rtData.logging > 2) debug msg, rtData
-        } catch(all) {
-            error "Error while executing physical command $device.$command($params):", rtData, null, all
-        }
-        if (rtData.piston.o?.ced) {
-            pause(rtData.piston.o.ced)
-            if (rtData.logging > 2) debug "Injected a ${rtData.piston.o.ced}ms delay after [$device].$command(${params ? "$params" : ''})", rtData
-        }
-    }
+		//we're using schedules instead
+		def statement = rtData.currentAction
+		def cs = [] + ((statement.tcp == 'b') || (statement.tcp == 'c') ? (rtData.stack?.cs ?: []) : [])
+		def ps = (statement.tcp == 'b') || (statement.tcp == 'p') ? 1 : 0
+		cs.removeAll{ it == 0 }
+		def schedule = [
+			t: now() + delay,
+			s: statement.$,
+			i: -3,
+			cs: cs,
+			ps: ps,
+			d: [
+				d: scheduleDevice,
+				c: command,
+				p: params
+			]
+		]
+		rtData.schedules.push(schedule)
+	} else {
+	try {
+		params = (params instanceof List) ? params : (params != null ? [params] : [])
+		//cleanup the params so that SONOS works
+		while (params.size() && (params[params.size()-1] == null)) params.pop()
+		def msg = timer ""
+		def skip = false
+		if (!rtData.piston.o?.dco && !disableCommandOptimization && !(command in ['setColorTemperature', 'setColor', 'setHue', 'setSaturation'])) {
+			def cmd = rtData.commands.physical[command]
+			if (cmd && cmd.a) {
+				if (cmd.v && !params.size()) {
+					//commands with no parameter that set an attribute to a preset value
+					if (getDeviceAttributeValue(rtData, device, cmd.a) == cmd.v) {
+						skip = true
+					}
+				} else if (params.size() == 1) {
+					if (getDeviceAttributeValue(rtData, device, cmd.a) == params[0]) {
+						skip = (command in ['setLevel', 'setInfraredLevel'] ? getDeviceAttributeValue(rtData, device, 'switch') == 'on' : true)
+					}
+				}
+			}
+		}
+		//if we're skipping, we already have a message
+		if (skip) {
+			msg.m = "Skipped execution of physical command [${device.label ?: device.name}].$command($params) because it would make no change to the device."
+		} else {
+			//def myres = runWithDelay(device, command, params, delay)
+			if (params.size()) {
+				//if (myres && delay) { //not supported in hubitat
+				if (delay) { //not supported in hubitat
+					pause(delay)
```

Comment:
No need to rewrite this code for now. I already am working around it in the first line of the method by using a schedule rather than a delay. There also isn't any checking on the delay time since it could be any number and not affect the piston execution time since it was run asynchronously if I'm remembering right. We don't want to lock the thread up here with a long pause. It could be improved with short pauses depending on how delayed the command is to offset short delays vs spinup time it would take to rerun the piston to get to this point again.

Reply:
I wondered on this. I did not pick out sensitivity to long vs. short here.   I agree if you knew a long pause with a schedule.  I thought all delays for device commands are short by rule in ST.